services/src/main/java/org/keycloak/organization/admin/resource/OrganizationIdentityProvidersResource.java

```diff
@@ -0,0 +1,139 @@
+
+@Provider
+public class OrganizationIdentityProvidersResource {
+
+    private final KeycloakSession session;
+    private final RealmModel realm;
+    private final OrganizationProvider organizationProvider;
+    private final OrganizationModel organization;
+    private final AdminPermissionEvaluator auth;
+    private final AdminEventBuilder adminEvent;
+
+    public OrganizationIdentityProvidersResource() {
+        // needed for registering to the JAX-RS stack
+        this(null, null, null, null);
+    }
+
+    public OrganizationIdentityProvidersResource(KeycloakSession session, OrganizationModel organization, AdminPermissionEvaluator auth, AdminEventBuilder adminEvent) {
+        this.session = session;
+        this.realm = session == null ? null : session.getContext().getRealm();
+        this.organizationProvider = session == null ? null : session.getProvider(OrganizationProvider.class);
+        this.organization = organization;
+        this.auth = auth;
+        this.adminEvent = adminEvent;
+    }
+
+    @POST
+    @Consumes(MediaType.APPLICATION_JSON)
+    public Response addIdentityProvider(IdentityProviderRepresentation providerRep) {
+        auth.realm().requireManageRealm();
+
+        Response response = new IdentityProvidersResource(realm, session, auth, adminEvent).create(providerRep);
+
+        if (Status.CREATED.getStatusCode() == response.getStatus()) {
+            try {
+                IdentityProviderModel identityProvider = realm.getIdentityProviderByAlias(providerRep.getAlias());
+
+                if (organizationProvider.addIdentityProvider(organization, identityProvider)) {
+                    return response;
+                }
+
+                throw ErrorResponse.error("Identity provider already associated to the organization", Status.BAD_REQUEST);
+            } catch (ModelException me) {
+                throw ErrorResponse.error(me.getMessage(), Status.BAD_REQUEST);
+            }
+        }
+
+        return response;
+    }
+
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    public List<IdentityProviderRepresentation> getIdentityProviders() {
+        auth.realm().requireManageRealm();
+        return organization.getIdentityProviders()
+                .map(this::toRepresentation)
+                .toList();
+    }
+
+    @Path("{alias}")
+    public IdentityProviderResource getIdentityProvider(@PathParam("alias") String alias) {
+        IdentityProviderModel broker = realm.getIdentityProviderByAlias(alias);
+        return new IdentityProviderResource(auth, realm, session, broker, adminEvent) {
+            @Override
+            public Response delete() {
+                Response response = super.delete();
+
+                if (organizationProvider.removeIdentityProvider(organization, broker)) {
+                    return response;
+                }
+
+                throw ErrorResponse.error("Identity provider not associated with the organization", Status.BAD_REQUEST);
```

Comment:
I might be overlooking something, but shouldn't there be different the error message? If I understand it correctly in this point there was something wrong during removing IdP from organization.

Reply:
If not removed is because it wasn't associated. Looks correct to me.

Reply:
This is the problem with methods like `removeIdentityProvider` that return a status boolean.  As the caller of the method, you don't actually know what went wrong.  Line 116 is assuming why the error happened.   And for now it's correct.  But later someone comes along to change that method and returns `false` for some other reason.    Then then assumption made in line 116 could be wrong.

I would suggest throwing an exception from inside the `removeIdentityProvider` method instead of using a status boolean.

Also, at some point we need to localize error messages that could end up in the UI.  But that may be a task for another day.